### ISSUE TITLE
Seed random number generator

### ISSUE THREAD

**Issue by emitra17** [lanl/PyBNF#31]
Allow specification of seed in config so run is reproducible

**Comment by wshlavacek** [comment]
## Implementation Plan

### Overview

Add a `random_seed` config key that seeds `np.random` at startup, making runs reproducible. All algorithmic randomness already flows through `np.random.*` on the main thread, so a single seed call is sufficient for the PyBNF side. Stochastic simulators (BioNetGen/NFSim) need separate handling.

### Step 1: Add `random_seed` config key

In `config.py`:
- Add `random_seed` to the set of recognized keys (integer, optional, default `None`)
- Validate it's a non-negative integer if provided

### Step 2: Seed `np.random` at startup

In `pybnf.py`, early in the run (before algorithm construction):
- If `random_seed` is set, call `np.random.seed(random_seed)`
- If `random_seed` is not set, generate a seed from system entropy via `np.random.randint(0, 2**31)`, call `np.random.seed()` with it, and log it so the user can reproduce after the fact
- Log the effective seed at INFO level: `"Random seed: %d"`

This is sufficient because:
- All random draws (parameter sampling, algorithm proposals, acceptance decisions, latin hypercube, bootstrap) use `np.random.*`
- All random draws happen on the main thread — the Dask workers only run simulators and don't call `np.random`

### Step 3: Handle job return order (the hard part)

The main run loop (`Algorithm.run()`) processes results via `custom_as_completed()`, meaning `got_result()` is called in whatever order simulations finish. For algorithms where `got_result()` triggers random draws (proposing new candidates), the sequence of random numbers consumed depends on return order.

**Synchronous/generation-based algorithms** (PSO, island DE, Scatter Search, Simplex, DREAM, all Bayesian MCMCs): These wait for all jobs in a generation/iteration to complete before proposing new candidates. The `got_result()` calls that trigger new random draws happen only after all results for the current batch are in. Within a batch, results are just stored — no random numbers are drawn. **These should be reproducible with just a seed.**

However, verify this claim for each algorithm by auditing whether `got_result()` draws random numbers before the generation boundary (e.g. for bookkeeping). If any do, the draws could be order-dependent.

**Asynchronous DE** (`ade`): Each `got_result()` immediately proposes a new candidate using `np.random.choice` and `np.random.random`. The order results arrive determines which random numbers go to which proposals. **This algorithm cannot be made reproducible without deeper changes** (e.g., per-index RNG streams or deterministic result ordering). Document this limitation.

### Step 4: Pass seeds to stochastic simulators (optional, separate PR)

For full reproducibility with stochastic simulators:
- When `random_seed` is set, use `np.random.randint()` to generate a per-simulation seed
- For BNGL: inject `seed=>VALUE` into the action command (PyBNF already detects and parses `seed=>` — see `BNGLModel.seeded`)
- For NFSim: similar mechanism via command-line `-seed` flag
- SBML/libRoadRunner: deterministic ODE solvers, no action needed

This is the most invasive part and could be a follow-up PR. Without it, the algorithmic trajectory is reproducible but stochastic simulation outputs may differ.

### Step 5: Documentation

- Document `random_seed` in the config reference
- Note the `ade` limitation
- Note that stochastic simulator reproducibility requires Step 4
- Note that the logged seed can be used to reproduce any run after the fact

### Step 6: Tests

Add tests that:
1. Set a seed, run `random_pset()` twice, verify identical results
2. Set a seed, run `random_latin_hypercube_psets()` twice, verify identical results
3. Set a seed, run 2-3 iterations of a synchronous algorithm (e.g. PSO), verify identical proposal sequences
4. Verify the config parser accepts and validates `random_seed`
5. Verify the seed is logged

---

## Pre-implementation test needs

Before starting this work, we should add **baseline characterization tests** that exercise the random→result pipeline for each algorithm. Currently, the algorithm tests (`test_particle_swarm.py`, `test_diff_evolution.py`, etc.) verify that iterations run without crashing and produce reasonable outputs, but they use **loose statistical assertions** like "some but not all proposals were accepted" or "values are within bounds."

These tests won't catch a subtle bug where seeding *appears* to work but the random stream is actually consumed in a nondeterministic order. We need tests that:

1. **Verify generation-boundary determinism for each synchronous algorithm**: Set a seed, feed results to `got_result()` in two different orders (simulating different job completion orders within a generation), and confirm the *next generation's proposals* are identical regardless. This is the key invariant that makes seeding work, and it's currently untested.

2. **Verify `ade` is order-dependent**: Same test as above but for async DE, confirming proposals *do* differ — documenting the known limitation.

These tests validate the assumptions the seed feature relies on. Without them, we could implement seeding and have it silently break for certain algorithms without any test catching it.

**Comment by wshlavacek** [comment]
Implemented issue #31.

Summary:
- Added a random_seed config key with parsing and validation.
- Seeded NumPy at PyBNF startup; when random_seed is omitted, PyBNF now generates an effective seed, logs it, and stores it in the runtime config.
- Updated user documentation in docs/config_keys.rst and the changelog so users know about the feature and its reproducibility limits.
- Added parser/config/startup tests for the new key.

Validation: .venv/bin/python -m pytest tests -q -> 395 passed.

Note: stochastic simulator-level reproducibility and result-order-dependent parallel algorithms remain documented limitations.
